apps/data-pipeline/sample-program-scraper/src/index.ts

```diff
@@ -0,0 +1,715 @@
+                : { type: "unknown" as const, value: code };
+            }),
+          ),
+          Promise.all(
+            year.winter.map(async (code) => {
+              const id = await lookupCourseId(db, code);
+              return id
+                ? { type: "courseId" as const, value: id }
+                : { type: "unknown" as const, value: code };
+            }),
+          ),
+          Promise.all(
+            year.spring.map(async (code) => {
+              const id = await lookupCourseId(db, code);
+              return id
+                ? { type: "courseId" as const, value: id }
+                : { type: "unknown" as const, value: code };
+            }),
+          ),
+        ]);
+
+        transformedSampleProgram.push({ year: year.year, fall, winter, spring });
+      }
+
+      transformedVariations.push({
+        label: variation.label,
+        sampleProgram: transformedSampleProgram,
+        variationNotes: variation.variationNotes,
+      });
+    }
+
+    transformed.push({
+      majorId: program.majorId,
+      programName: program.programName,
+      variations: transformedVariations,
+    });
+  }
+
+  logger.info("Course code transformation complete");
+  return transformed;
+}
+
+async function storeSampleProgramsInDB(
+  db: ReturnType<typeof database>,
+  scrapedPrograms: {
+    majorId: string;
+    programName: string;
+    variations: {
+      label?: string;
+      sampleProgram: UnvalidatedSampleProgramEntry[];
+      variationNotes: string[];
+    }[];
+  }[],
+) {
+  if (!scrapedPrograms.length) {
+    logger.info("No sample programs to store.");
+    return;
+  }
+
+  // Transform course codes to IDs
+  const transformedPrograms = await transformCourseCodesToIds(db, scrapedPrograms);
+
+  logger.info("Fetching sample programs from database...");
+
+  const catalogueRows = transformedPrograms.map((program) => ({
+    id: program.majorId,
+    programName: program.programName,
+  }));
+
+  // Prepare child table rows (sample_program_variation)
+  const variationRows = transformedPrograms.flatMap((program) =>
+    program.variations.map((variation) => ({
+      programId: program.majorId,
+      label: orNull(variation.label),
+      sampleProgram: variation.sampleProgram,
+      variationNotes: variation.variationNotes,
+    })),
+  );
+
+  const programIds = catalogueRows.map((r) => r.id);
+
+  const existingCatalogPrograms = await db
+    .select()
+    .from(catalogProgram)
+    .where(inArray(catalogProgram.id, programIds));
+
+  const existingVariations = await db
+    .select()
+    .from(sampleProgramVariation)
+    .where(inArray(sampleProgramVariation.programId, programIds));
+
+  // Combine into unified structure for comparison
+  const dbData = { catalogPrograms: existingCatalogPrograms, variations: existingVariations };
+  const scrapedData = { catalogPrograms: catalogueRows, variations: variationRows };
+
+  // Sort both structures
```

Comment:
I think the variable name and actual code make it clear that this is a sort, so this comment is not needed.